### ISSUE TITLE
Split selection

### ISSUE THREAD

**Issue by Kaligule** [magnars/multiple-cursors.el#292]
It would be cool if we could split a region (between mark and point) into multiple selections, each with its own cursor (The idea is from [kakoune](http://vimeo.com/82711574)). For example you could have the following text:

> {one, two, three, four, five}

Where the mark might be after { and point might be before }, so the region is _one, two, three, four, five_. Then after calling mc/split-region you would be promted for a split string and we would enter ", ". This would then result in 5 cursors, each one with an active region. Every cursor would be at the end of a word.

**Comment by nispio** [comment]
It's not perfect, but I will point out that you can already get most of the way there with `mc/mark-all-in-region-regexp`

**Comment by nispio** [comment]
Here is a set of functions that should do what you are after:
```
(defun mc/split-region (arg beg end)
  "Split the current region into sub-regions at each occurrence of a given string"
  (interactive "p\nr")
  (let ((search (read-string "Split region on string: "))
        (inc (cond
              ((< 1 arg) 'after)
              ((> 0 arg) 'before)
              (t nil))))
    (mc--do-split-region (regexp-quote search) beg end inc)))

(defun mc/split-region-regexp (arg beg end)
  "Split the current region into sub-regions at each occurrence of a given regexp"
  (interactive "p\nr")
  (let ((regexp (read-regexp "Split region on regexp: "))
        (inc (cond
              ((< 1 arg) 'after)
              ((> 0 arg) 'before)
              (t nil))))
    (mc--do-split-region regexp beg end inc)))

(defun mc--do-split-region (regexp beg end &optional include-match)
  (if (string= regexp "")
      (message "Mark aborted")
    (let ((lastmatch beg)
          (case-fold-search nil))
      (mc/remove-fake-cursors)
      (deactivate-mark)
      (goto-char beg)
      (while (and (< (point) end) ; can happen because of (forward-char)
                  (search-forward-regexp regexp end t))
        (if (eq include-match 'after)
            (goto-char (match-end 0))
          (goto-char (match-beginning 0)))
        (push-mark lastmatch t nil)
        (mc/create-fake-cursor-at-point)
        (if (eq include-match 'before)
            (goto-char (match-beginning 0))
          (goto-char (match-end 0)))
        (setq lastmatch (point))
        (goto-char (match-end 0))
        (when (= (point) (match-beginning 0))
          (forward-char)))
      (if (not lastmatch)
          (multiple-cursors-mode 0)
        (goto-char end)
        (push-mark lastmatch t nil)
        (multiple-cursors-mode 1)))))
```

Let me know what you think, and if it works well I will submit a PR

**Comment by Kaligule** [comment]
This is great, thank you. (I am learning to read emacs-lisp at the moment at this is quite helpfull).
The idea to use `mc/mark-all-in-region-regexp` is clever, I didn't see that immediatelly.

It doesn't seem to do exactly the right thing, though. I tried to mark a sentence ("Everybody should love emacs.") and ran `mc/split-region` with space as an argument. I got 4 cursors (great, what I wished for) but there was no region associated with any of them. I had expected each cursor to have a region with exactly one word in it.

This way the region would be really _split_ into multiple smaller regions, not just multiple cursors.

**Comment by nispio** [comment]
I will look into this.  The code was working as expected when I wrote it (emacs-25.1, CentOS7, latest multiple-cursors), but I didn't try running with `emacs -q` to see if anything in my config was affecting behavior.  I will test it more thoroughly and see if I can reproduce your problem.

**Comment by nispio** [comment]
I found the issue.  Simply delete the line `(deactivate-mark)` from the `mc--do-split-region` function.  I have edited the code above so that it should work as-written.  Test it out again, if you don't mind, and let me know how it works.

I haven't written a proper docstring for the functions yet, but giving a prefix argument should cause the "splitting string" (e.g. the comma or space) to be included at the end of the newly selected regions.  The default behavior is to exclude it from the selection.  A negative prefix argument causes the splitting string to be included at the beginning of the newly selected regions.

```
[one],[two],[three]   ; No prefix argument excludes the comma from the new regions
[one,][two,][three]   ; A prefix argument causes the comma to be included after the text
[one][,two][,three]   ; A negative prefix argument causes the comma to be included before the text
```
